File: app/src/main/java/com/bing/example/main/adapter/VideosAdapter.java
package com.bing.example.main.adapter;

import android.view.View;
import android.view.ViewGroup;

import com.bing.example.R;
import com.bing.example.databinding.ItemVideoBinding;
import com.bing.example.model.entity.VideoInfo;
import com.bing.mvvmbase.base.recycleview.BaseRecycleViewAdapter;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.TreeSet;

import androidx.annotation.NonNull;

public class VideosAdapter extends BaseRecycleViewAdapter<VideoInfo, VideoInfoHolder> {
        public static final int MODE_NORMAL = 0;
        public static final int MODE_SELECT = 1;

        private int mCurrentMode = MODE_NORMAL;

	private OnClickListener mListener;
	private OnModeChangeListener mModeChangeListener;

        private Set<Integer> mSelectedPostions = new TreeSet<>();

        public Set<Integer> getSelectedPostions() {
                return mSelectedPostions;
        }

        public List<VideoInfo> getSelectedData() {
                List<VideoInfo> infos = new ArrayList<>();
                for (Integer position : mSelectedPostions) {
                        infos.add(mData.get(position));
                }
                return infos;
        }

        public VideosAdapter(OnClickListener listener, OnModeChangeListener modeChangeListener) {
		mListener = listener;
		mModeChangeListener = modeChangeListener;
	}

        @Override
        protected VideoInfoHolder createHolder(@NonNull ViewGroup viewGroup, int i) {
                return new VideoInfoHolder((ItemVideoBinding)getBinding(viewGroup, i, R.layout.item_video));
        }

        @Override
        protected void bindData(VideoInfoHolder holder, int position) {
                ItemVideoBinding binding = holder.getBinding();
                binding.setVideoInfo(mData.get(position));
                binding.setMode(mCurrentMode);
                binding.setSelected(mSelectedPostions.contains(position));
                binding.backContent.setOnLongClickListener(v -> {
                        intoMutiSelectMode(position);
                        return true;
                });
                binding.thumb.setOnLongClickListener(v -> {
                        intoMutiSelectMode(position);
                        return true;
                });
                binding.rename.setOnLongClickListener(v -> {
                        intoMutiSelectMode(position);
                        return true;
                });
                binding.delete.setOnLongClickListener(v -> {
                        intoMutiSelectMode(position);
                        return true;
                });
                binding.share.setOnLongClickListener(v -> {
                        intoMutiSelectMode(position);
                        return true;
                });
                binding.thumb.setOnClickListener(v -> {
                        if (mListener != null) {
                                mListener.onClickImage(position, mData.get(position));
                        }
                });
	        binding.rename.setOnClickListener(v -> {
		        if (mListener != null) {
			        mListener.onClickRename(position, mData.get(position));
		        }
	        });
	        binding.delete.setOnClickListener(v -> {
		        if (mListener != null) {
			        mListener.onClickDelete(position, mData.get(position));
		        }
	        });
	        binding.share.setOnClickListener(v -> {
		        if (mListener != null) {
			        mListener.onClickShare(position, mData.get(position));
		        }
	        });
                binding.foreground.setOnClickListener(v -> {
			onClickForeground(position);
                });
                binding.checkbox.setOnClickListener(v -> onClickForeground(position));
        }

        private void onClickForeground(int position) {
	        if (mSelectedPostions.contains(position)) {
		        mSelectedPostions.remove(position);
		        notifyItemChanged(position);
	        } else {
		        mSelectedPostions.add(position);
		        notifyItemChanged(position);
	        }
	        if (mListener != null) {
		        mListener.onClickFore(position, mSelectedPostions.size());
	        }
        }

        private void intoMutiSelectMode(int position) {
                mCurrentMode = MODE_SELECT;
                mSelectedPostions.add(position);
                if (mModeChangeListener != null) {
                        mModeChangeListener.onModeTo(MODE_SELECT);
                }
                notifyDataSetChanged();
        }

        public void intoNormalMode() {
                mCurrentMode = MODE_NORMAL;
                if (mModeChangeListener != null) {
                        mModeChangeListener.onModeTo(MODE_NORMAL);
                }
                mSelectedPostions.clear();
                notifyDataSetChanged();
        }

	public interface OnClickListener {
                void onClickImage(int position, VideoInfo videoInfo);
                void onClickRename(int position, VideoInfo videoInfo);
                void onClickDelete(int position, VideoInfo videoInfo);
                void onClickShare(int position, VideoInfo videoInfo);
                void onClickFore(int position, int selectedNumber);
	}

	public interface OnModeChangeListener {
                void onModeTo(int mode);
        }
}

File: app/src/main/java/com/bing/example/main/videolist/VideosAdapter.kt
package com.bing.example.main.videolist

import android.view.ViewGroup

import com.bing.example.R
import com.bing.example.databinding.ItemVideoBinding
import com.bing.example.model.entity.VideoInfo
import com.bing.mvvmbase.base.recycleview.BaseRecycleViewAdapter

import java.util.ArrayList
import java.util.TreeSet

class VideosAdapter(private val mListener: OnClickListener?, private val mModeChangeListener: OnModeChangeListener?) : BaseRecycleViewAdapter<VideoInfo, VideoInfoHolder>() {

        var currentMode = MODE_NORMAL

        val selectedPostions = TreeSet<Int>()

        val selectedData: List<VideoInfo>
                get() {
                        val infos = ArrayList<VideoInfo>()
                        if (data != null) {
                                for (position in selectedPostions) {
                                        infos.add(data!![position])
                                }
                        }
                        return infos
                }

        override fun createHolder(parent: ViewGroup, viewType: Int): VideoInfoHolder {
                return VideoInfoHolder(getBinding(parent, viewType, R.layout.item_video) as ItemVideoBinding)
        }

        override fun bindData(holder: VideoInfoHolder, position: Int) {
                val binding = holder.binding
                val videoInfo = data!![position]
                binding.videoInfo = videoInfo
                binding.mode = currentMode
                binding.selected = selectedPostions.contains(position)
                binding.backContent.setOnLongClickListener {
                        intoMutiSelectMode(position)
                        true
                }
                binding.thumb.setOnLongClickListener {
                        intoMutiSelectMode(position)
                        true
                }
                binding.rename.setOnLongClickListener {
                        intoMutiSelectMode(position)
                        true
                }
                binding.delete.setOnLongClickListener {
                        intoMutiSelectMode(position)
                        true
                }
                binding.share.setOnLongClickListener {
                        intoMutiSelectMode(position)
                        true
                }
                binding.thumb.setOnClickListener {
                        mListener?.onClickImage(position, videoInfo)
                }
                binding.rename.setOnClickListener {
                        mListener?.onClickRename(position, videoInfo)
                }
                binding.delete.setOnClickListener {
                        mListener?.onClickDelete(position, videoInfo)
                }
                binding.share.setOnClickListener {
                        mListener?.onClickShare(position, videoInfo)
                }
                binding.foreground.setOnClickListener { onClickForeground(position) }
                binding.checkbox.setOnClickListener { onClickForeground(position) }
        }

        private fun onClickForeground(position: Int) {
                if (selectedPostions.contains(position)) {
                        selectedPostions.remove(position)
                        notifyItemChanged(position)
                } else {
                        selectedPostions.add(position)
                        notifyItemChanged(position)
                }
                mListener?.onClickFore(position, selectedPostions.size)
        }

        private fun intoMutiSelectMode(position: Int) {
                currentMode = MODE_SELECT
                selectedPostions.add(position)
                mModeChangeListener?.onModeTo(MODE_SELECT)
                notifyDataSetChanged()
        }

        fun intoNormalMode() {
                currentMode = MODE_NORMAL
                mModeChangeListener?.onModeTo(MODE_NORMAL)
                selectedPostions.clear()
                notifyDataSetChanged()
        }

        interface OnClickListener {
                fun onClickImage(position: Int, videoInfo: VideoInfo)
                fun onClickRename(position: Int, videoInfo: VideoInfo)
                fun onClickDelete(position: Int, videoInfo: VideoInfo)
                fun onClickShare(position: Int, videoInfo: VideoInfo)
                fun onClickFore(position: Int, selectedNumber: Int)
        }

        interface OnModeChangeListener {
                fun onModeTo(mode: Int)
        }

        companion object {
                const val MODE_NORMAL = 0
                const val MODE_SELECT = 1
        }
}
